objc/include/onnxruntime/ort_session.h

```diff
@@ -0,0 +1,46 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#import <Foundation/Foundation.h>
+
+@class ORTEnv;
+@class ORTValue;
+
+// TODO
+@interface ORTSessionOptions : NSObject
+@end
+
+// TODO
+@interface ORTRunOptions : NSObject
+@end
+
+/**
+ * An ORT session loads and runs a model.
+ */
+@interface ORTSession : NSObject
+
+/**
+ * Creates an ORT Session.
+ *
+ * @param env The ORT Environment instance.
+ * @param path The path to the ONNX model.
+ * @param error Optional error information set if an error occurs.
+ * @return The instance, or nil if an error occurs.
+ */
+- (instancetype)initWithEnv:(ORTEnv*)env
+                  modelPath:(NSString*)path
+                      error:(NSError**)error;
+
+/**
+ * Runs the model with pre-allocated inputs and outputs.
+ *
+ * @param inputs Dictionary of input names to input ORT values.
+ * @param outputs Dictionary of output names to output ORT values.
+ * @param error Optional error information set if an error occurs.
+ * @return Whether the model was run successfully.
+ */
+- (BOOL)runWithInputs:(NSDictionary<NSString*, ORTValue*>*)inputs
+              outputs:(NSDictionary<NSString*, ORTValue*>*)outputs
```

Comment:
Memory management for the outputs isn't clear to me. The user may or may not provide a pre-allocated output. Do they just init the `outputs` dictionary with nulls if not pre-allocating?

Reply:
for this rudimentary initial API, all inputs and outputs should be pre-allocated. I can update the description to say "pre-allocated outputs".
the name of the function is "runWithInputs:outputs:error:", so "WithInputs" seemed ok, but open to suggestions.

---
In reply to: [616198922](https://github.com/microsoft/onnxruntime/pull/7366#discussion_r616198922) [](ancestors = 616198922)

Reply:
Ah ok - not that familiar with objective-c and how the method name is a combination of the input names. 

Would be better to start with outputs not being pre-allocated. For models that produce outputs of a dynamic size it's not possible to preallocate (e.g. number of bounding boxes could be input dependent).

If you want to keep this variant I'd suggest using 'preallocatedOutputs' to clarify unless support for non-pre-allocated outputs will be added via this method. 

---
In reply to: [616233216](https://github.com/microsoft/onnxruntime/pull/7366#discussion_r616233216) [](ancestors = 616233216,616198922)

Reply:
ok. I can add it - it will require supporting ORT values allocated by ORT first though.

---
In reply to: [616258034](https://github.com/microsoft/onnxruntime/pull/7366#discussion_r616258034) [](ancestors = 616258034,616233216,616198922)